alf/layers.py

```diff
@@ -110,6 +112,128 @@ def forward(self, input):
             input, num_classes=self._num_classes).to(torch.float32)
 
 
+@gin.configurable
+class FixedDecodingLayer(nn.Module):
+    def __init__(self,
+                 input_size,
+                 output_size,
+                 basis_type="rbf",
+                 sigma=1.,
+                 tau=0.5):
+        """A layer that uses a set of fixed basis for decoding the inputs.
+        Args:
+            input_size (int): the size of input to be decoded, representing the
+                number of representation coefficients
+            output_size (int): the size of the decoded output
+            basis_type (str): the type of basis to be used for decoding
+                - "poly": polynomial basis using Vandermonde matrix
+                - "cheb": polynomial basis using Chebyshev polynomials
+                - "rbf": radial basis functions
+                - "haar": haar wavelet basis
+            sigma (float): the bandwidth parameter used for RBF basis.
+                If None, a default value of 1. will be used.
+            tau (float): a factor for weighting the basis exponentially
+                according to the order (``n``) of the basis, i.e., ``tau**n```
+        """
+        # get the argument list with vals
+        self._kwargs = copy.deepcopy(locals())
+        self._kwargs.pop('self')
+        self._kwargs.pop('__class__')
+
+        super(FixedDecodingLayer, self).__init__()
+
+        assert input_size > 0, "input_size should be at least one"
+        assert basis_type in {"poly", "cheb", "rbf", "haar"
+                              }, ("the specified method "
+                                  "{} is not supported".format(basis_type))
+
+        self._B = nn.Linear(input_size, output_size, bias=False)
+
+        def _polyvander_matrix(n, D, tau=tau):
+            # non-square matrix [n, D + 1]
+            x = torch.linspace(-1, 1, n)
+            B = torch.empty(n, D + 1)
+            for d in range(D + 1):
+                B[:, d] = x**d
+            # weight for encoding the preference to low-frequency basis
+            exp_factor = torch.arange(D + 1).float()
+            basis_weight = tau**exp_factor
+            return B * basis_weight
+
+        def _chebvander_matrix(n, D, tau=tau):
+            # non-square matrix [n, D + 1]
+            x = np.linspace(-1, 1, n)
+            B = torch.from_numpy(np.polynomial.chebyshev.chebvander(x, D)).to(
```

Comment:
as_tensor() can convert to the default device by default.

Reply:
Changed to use as_tensor().